File: python_system/python_shared/api/user_profiling.py
"""
user_profiling.py - User Profiling API
Track and analyze user preferences and behavior
Converted from PHP to Python
"""

from fastapi import APIRouter, Request, Query
from fastapi.responses import JSONResponse
from pathlib import Path
from datetime import datetime, timedelta
import json
import time
import hashlib

router = APIRouter()

PROFILES_PATH = Path(__file__).parent.parent / "data" / "user_profiles"
PROFILES_PATH.mkdir(parents=True, exist_ok=True)


def get_device_id(request: Request) -> str:
    """Generate a unique device ID from request"""
    ip = request.client.host if request.client else "unknown"
    ua = request.headers.get("user-agent", "unknown")
    combined = f"{ip}:{ua}"
    return hashlib.md5(combined.encode()).hexdigest()[:16]


@router.post("/user_profile")
async def update_user_profile(request: Request):
    """Update user profile with interaction data"""
    try:
        data = await request.json()

        device_id = data.get("device_id") or get_device_id(request)
        ad_id = data.get("ad_id", "")
        action = data.get("action", "")  # view, like, favorite, contact, click
        category = data.get("category", "")
        duration = data.get("duration", 0)

        # Load or create profile
        profile_file = PROFILES_PATH / f"{device_id}.json"
        profile = {
            "device_id": device_id,
            "created_at": datetime.now().isoformat(),
            "interests": {},
            "category_affinity": {},
            "interaction_history": [],
            "total_views": 0,
            "total_likes": 0,
            "total_favorites": 0,
            "total_contacts": 0,
            "avg_view_duration": 0
        }

        if profile_file.exists():
            with open(profile_file, "r") as f:
                stored = json.load(f)
                profile.update(stored)

        # Update profile based on action
        timestamp = int(time.time())

        if action == "view":
            profile["total_views"] += 1
            if category:
                profile["category_affinity"][category] = profile["category_affinity"].get(category, 0) + 1
        elif action == "like":
            profile["total_likes"] += 1
            if category:
                profile["category_affinity"][category] = profile["category_affinity"].get(category, 0) + 3
        elif action == "favorite":
            profile["total_favorites"] += 1
            if category:
                profile["category_affinity"][category] = profile["category_affinity"].get(category, 0) + 5
        elif action == "contact":
            profile["total_contacts"] += 1
            if category:
                profile["category_affinity"][category] = profile["category_affinity"].get(category, 0) + 10

        # Track duration
        if duration > 0:
            durations = [h.get("duration", 0) for h in profile["interaction_history"] if h.get("duration")]
            durations.append(duration)
            profile["avg_view_duration"] = round(sum(durations) / len(durations), 2)

        # Add to history (keep last 100)
        profile["interaction_history"].append({
            "ad_id": ad_id,
            "action": action,
            "category": category,
            "duration": duration,
            "timestamp": timestamp
        })
        profile["interaction_history"] = profile["interaction_history"][-100:]

        # Update timestamps
        profile["last_active"] = datetime.now().isoformat()
        profile["updated_at"] = datetime.now().isoformat()

        # Save profile
        with open(profile_file, "w") as f:
            json.dump(profile, f, indent=2)

        return {
            "success": True,
            "device_id": device_id,
            "profile_updated": True
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )


@router.get("/user_profile")
async def get_user_profile(request: Request, device_id: str = None):
    """Get user profile and recommendations"""
    try:
        if not device_id:
            device_id = get_device_id(request)

        profile_file = PROFILES_PATH / f"{device_id}.json"

        if not profile_file.exists():
            return {
                "success": True,
                "profile": None,
                "message": "No profile found"
            }

        with open(profile_file, "r") as f:
            profile = json.load(f)

        # Generate category recommendations
        recommendations = []
        if profile.get("category_affinity"):
            sorted_categories = sorted(
                profile["category_affinity"].items(),
                key=lambda x: x[1],
                reverse=True
            )
            recommendations = [cat for cat, _ in sorted_categories[:5]]

        return {
            "success": True,
            "profile": {
                "device_id": profile["device_id"],
                "total_views": profile.get("total_views", 0),
                "total_likes": profile.get("total_likes", 0),
                "total_favorites": profile.get("total_favorites", 0),
                "avg_view_duration": profile.get("avg_view_duration", 0),
                "category_affinity": profile.get("category_affinity", {}),
                "last_active": profile.get("last_active")
            },
            "recommended_categories": recommendations
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )


@router.get("/user_profile/preferences")
async def get_preferences(request: Request, device_id: str = None):
    """Get user preferences for ad targeting"""
    try:
        if not device_id:
            device_id = get_device_id(request)

        profile_file = PROFILES_PATH / f"{device_id}.json"

        if not profile_file.exists():
            return {"success": True, "preferences": {}, "has_profile": False}

        with open(profile_file, "r") as f:
            profile = json.load(f)

        # Calculate preferences
        category_affinity = profile.get("category_affinity", {})
        total_affinity = sum(category_affinity.values()) or 1

        preferences = {
            cat: round(score / total_affinity, 3)
            for cat, score in category_affinity.items()
        }

        return {
            "success": True,
            "has_profile": True,
            "preferences": preferences,
            "engagement_score": min(100, profile.get("total_views", 0) + profile.get("total_likes", 0) * 5)
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )


@router.delete("/user_profile")
async def delete_user_profile(request: Request, device_id: str = None):
    """Delete user profile (GDPR compliance)"""
    try:
        if not device_id:
            device_id = get_device_id(request)

        profile_file = PROFILES_PATH / f"{device_id}.json"

        if profile_file.exists():
            profile_file.unlink()
            return {"success": True, "message": "Profile deleted"}
        else:
            return {"success": True, "message": "No profile found"}

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )

File: python_system/python_shared/api/user_profiling.php
<?php
/********************************************
 * Advanced User Profiling API
 * Device fingerprinting, preference learning, AI recommendations
 ********************************************/
header("Content-Type: application/json");

// Get request data
$input = json_decode(file_get_contents('php://input'), true);
$action = $input['action'] ?? ''; // 'get_profile', 'update_profile', 'get_recommendations'
$deviceId = $input['device_id'] ?? '';
$preferences = $input['preferences'] ?? [];

if (empty($action)) {
    http_response_code(400);
    echo json_encode(['success' => false, 'message' => 'No action specified']);
    exit;
}

$profilesBase = __DIR__ . "/../companies/user_profiles/";
if (!is_dir($profilesBase)) {
    mkdir($profilesBase, 0775, true);
}

// ============================================
// GET USER PROFILE
// ============================================
if ($action === 'get_profile' && $deviceId) {
    $profileFile = "$profilesBase/$deviceId.json";

    if (file_exists($profileFile)) {
        $profile = json_decode(file_get_contents($profileFile), true);
        echo json_encode([
            'success' => true,
            'profile' => $profile,
            'is_new' => false
        ]);
    } else {
        // Create new profile
        $profile = [
            'device_id' => $deviceId,
            'created_at' => time(),
            'last_active' => time(),
            'total_sessions' => 1,
            'preferences' => [
                'liked_categories' => [],
                'disliked_categories' => [],
                'liked_ads' => [],
                'disliked_ads' => [],
                'favorite_ads' => [],
                'viewed_ads' => [],
                'contacted_ads' => []
            ],
            'behavior' => [
                'avg_time_per_ad' => 0,
                'total_time_spent' => 0,
                'total_interactions' => 0,
                'preferred_ad_types' => [], // image, video
                'browse_patterns' => [],
                'peak_activity_hours' => []
            ],
            'demographics' => [
                'device_type' => $input['device_type'] ?? 'unknown',
                'browser' => $input['browser'] ?? 'unknown',
                'os' => $input['os'] ?? 'unknown',
                'screen_size' => $input['screen_size'] ?? 'unknown',
                'timezone' => $input['timezone'] ?? 'unknown'
            ],
            'ml_scores' => [
                'engagement_score' => 0,
                'conversion_probability' => 0,
                'value_score' => 0
            ]
        ];

        file_put_contents($profileFile, json_encode($profile, JSON_PRETTY_PRINT));

        echo json_encode([
            'success' => true,
            'profile' => $profile,
            'is_new' => true
        ]);
    }
    exit;
}

// ============================================
// UPDATE USER PROFILE
// ============================================
if ($action === 'update_profile' && $deviceId && $preferences) {
    $profileFile = "$profilesBase/$deviceId.json";
    $profile = [];

    if (file_exists($profileFile)) {
        $profile = json_decode(file_get_contents($profileFile), true);
    } else {
        http_response_code(404);
        echo json_encode(['success' => false, 'message' => 'Profile not found']);
        exit;
    }

    // Update profile with new preferences
    $profile['last_active'] = time();
    $profile['total_sessions']++;

    // Update preferences
    foreach ($preferences as $key => $value) {
        if (isset($profile['preferences'][$key])) {
            if (is_array($value)) {
                $profile['preferences'][$key] = array_unique(array_merge($profile['preferences'][$key], $value));
                // Limit array size
                $profile['preferences'][$key] = array_slice($profile['preferences'][$key], -100);
            } else {
                $profile['preferences'][$key] = $value;
            }
        }
    }

    // Update behavior patterns
    if (isset($preferences['time_spent'])) {
        $profile['behavior']['total_time_spent'] += $preferences['time_spent'];
        $profile['behavior']['total_interactions']++;
        $profile['behavior']['avg_time_per_ad'] = round(
            $profile['behavior']['total_time_spent'] / $profile['behavior']['total_interactions'],
            2
        );
    }

    // Track activity hours
    $currentHour = (int)date('G');
    if (!isset($profile['behavior']['peak_activity_hours'][$currentHour])) {
        $profile['behavior']['peak_activity_hours'][$currentHour] = 0;
    }
    $profile['behavior']['peak_activity_hours'][$currentHour]++;

    // Calculate ML scores
    $profile['ml_scores'] = calculateMLScores($profile);

    file_put_contents($profileFile, json_encode($profile, JSON_PRETTY_PRINT));

    echo json_encode([
        'success' => true,
        'profile' => $profile,
        'message' => 'Profile updated'
    ]);
    exit;
}

// ============================================
// GET PERSONALIZED RECOMMENDATIONS
// ============================================
if ($action === 'get_recommendations' && $deviceId) {
    $profileFile = "$profilesBase/$deviceId.json";

    if (!file_exists($profileFile)) {
        echo json_encode([
            'success' => true,
            'recommendations' => [],
            'algorithm' => 'default',
            'message' => 'No profile found, showing default'
        ]);
        exit;
    }

    $profile = json_decode(file_get_contents($profileFile), true);

    // Load all ads
    $adsBase = __DIR__ . "/../companies/data/";
    $allAds = [];

    foreach (scandir($adsBase) as $category) {
        if ($category === '.' || $category === '..') continue;

        $categoryPath = "$adsBase/$category";
        if (!is_dir($categoryPath)) continue;

        foreach (scandir($categoryPath) as $company) {
            if ($company === '.' || $company === '..') continue;

            $companyPath = "$categoryPath/$company";
            if (!is_dir($companyPath)) continue;

            foreach (scandir($companyPath) as $adFolder) {
                if ($adFolder === '.' || $adFolder === '..') continue;

                $metaFile = "$companyPath/$adFolder/meta.json";
                if (!file_exists($metaFile)) continue;

                $meta = json_decode(file_get_contents($metaFile), true);
                if (!$meta) continue;

                // Calculate relevance score for this ad
                $score = calculateRelevanceScore($meta, $profile, $category);

                $allAds[] = [
                    'ad' => $meta,
                    'category' => $category,
                    'score' => $score
                ];
            }
        }
    }

    // Sort by relevance score
    usort($allAds, function($a, $b) {
        return $b['score'] - $a['score'];
    });

    // Get top recommendations
    $recommendations = array_slice($allAds, 0, 50);

    echo json_encode([
        'success' => true,
        'recommendations' => $recommendations,
        'algorithm' => 'ml_personalized',
        'profile_strength' => calculateProfileStrength($profile),
        'total_ads_scored' => count($allAds)
    ]);
    exit;
}

// ============================================
// ML SCORING FUNCTIONS
// ============================================
function calculateRelevanceScore($ad, $profile, $category) {
    $score = 0;
    $weights = [
        'favorite' => 40,           // Highest weight for favorites
        'category_match' => 25,
        'previous_likes' => 20,
        'time_engagement' => 15,
        'recency' => 10,
        'novelty' => 10,
        'popularity' => 5
    ];

    // Favorite ads (STRONGEST signal)
    $favoriteAds = $profile['preferences']['favorite_ads'] ?? [];
    if (in_array($ad['ad_id'], $favoriteAds)) {
        $score += $weights['favorite'];
    }

    // Category preference
    $likedCategories = $profile['preferences']['liked_categories'] ?? [];
    if (in_array($category, $likedCategories)) {
        $score += $weights['category_match'];
    }

    // Previous likes
    $likedAds = $profile['preferences']['liked_ads'] ?? [];
    $dislikedAds = $profile['preferences']['disliked_ads'] ?? [];

    if (in_array($ad['ad_id'], $likedAds)) {
        $score += $weights['previous_likes'];
    }
    if (in_array($ad['ad_id'], $dislikedAds)) {
        $score -= $weights['previous_likes'] * 2; // Heavy penalty
    }

    // Time engagement pattern
    $avgTime = $profile['behavior']['avg_time_per_ad'] ?? 0;
    if ($avgTime > 20) { // User engages deeply
        $score += $weights['time_engagement'];
    }

    // Recency (newer ads get slight boost)
    $adAge = time() - ($ad['timestamp'] ?? time());
    $daysSincePosted = $adAge / 86400;
    if ($daysSincePosted < 7) {
        $score += $weights['recency'] * (1 - ($daysSincePosted / 7));
    }

    // Novelty (haven't seen this ad before)
    $viewedAds = $profile['preferences']['viewed_ads'] ?? [];
    if (!in_array($ad['ad_id'], $viewedAds)) {
        $score += $weights['novelty'];
    }

    // Popularity boost
    $analyticsFile = __DIR__ . "/../companies/analytics/{$ad['ad_id']}.json";
    if (file_exists($analyticsFile)) {
        $analytics = json_decode(file_get_contents($analyticsFile), true);
        $likes = $analytics['total_likes'] ?? 0;
        $dislikes = $analytics['total_dislikes'] ?? 0;

        if ($likes > $dislikes && $likes > 10) {
            $score += $weights['popularity'];
        }
    }

    // Random factor for diversity (5%)
    $score += rand(0, 5);

    return $score;
}

function calculateMLScores($profile) {
    $scores = [
        'engagement_score' => 0,
        'conversion_probability' => 0,
        'value_score' => 0
    ];

    // Engagement score (0-100)
    $totalInteractions = $profile['behavior']['total_interactions'] ?? 0;
    $avgTime = $profile['behavior']['avg_time_per_ad'] ?? 0;
    $likedAds = count($profile['preferences']['liked_ads'] ?? []);

    $scores['engagement_score'] = min(100, (
        ($totalInteractions * 2) +
        ($avgTime * 1.5) +
        ($likedAds * 3)
    ));

    // Conversion probability (0-1)
    $contactedAds = count($profile['preferences']['contacted_ads'] ?? []);
    $viewedAds = count($profile['preferences']['viewed_ads'] ?? []);

    if ($viewedAds > 0) {
        $scores['conversion_probability'] = round($contactedAds / $viewedAds, 4);
    }

    // Value score (estimated user value)
    $scores['value_score'] = round(
        ($contactedAds * 5) +
        ($likedAds * 1) +
        ($scores['engagement_score'] * 0.1),
        2
    );

    return $scores;
}

function calculateProfileStrength($profile) {
    $totalInteractions = $profile['behavior']['total_interactions'] ?? 0;
    $sessions = $profile['total_sessions'] ?? 1;

    if ($totalInteractions < 5) return 'weak';
    if ($totalInteractions < 20) return 'moderate';
    if ($totalInteractions < 50) return 'strong';
    return 'very_strong';
}

echo json_encode(['success' => false, 'message' => 'Invalid action']);

